pyrobosim/pyrobosim/sensors/types.py

```diff
@@ -0,0 +1,88 @@
+"""
+Basic types for sensor simulation.
+"""
+
+from matplotlib.artist import Artist
+from threading import Thread
+from typing import Any
+
+
+class Sensor:
+    """
+    Generic sensor class that helps with type hinting.
+
+    When implementing a new sensor, you should subclass from this class.
+    """
+
+    def __init__(self) -> None:
+        from ..core.robot import Robot
+
+        self.robot: Robot | None = None
+        self.thread = Thread(target=self.thread_function)
+
+    def update(self) -> None:
+        """
+        Performs the sensor calculation.
+
+        Must be implemented in your sensor implementation.
+        """
+        raise NotImplementedError("Must implement update()!")
+
+    def get_measurement(self) -> Any:
+        """
+        Returns the latest measurement from the sensor.
+
+        Must be implemented in your sensor implementation.
+
+        :return: The latest measurement.
+        """
+        raise NotImplementedError("Must implement get_measurement()!")
+
+    def thread_function(self) -> None:
+        """
+        Defines the sensor update function to run in a background thread.
+
+        If the sensor has nothing to do in the background, you can leave this unimplemented in your sensor.
+        """
+        return
+
+    def start_thread(self) -> None:
+        """
+        Starts the thread defined by your sensor's implementation of `thread_function()`.
+        """
+        if self.thread is not None:
+            self.thread.start()
+
+    def stop_thread(self) -> None:
+        """
+        Stops the actively running sensor thread.
+
+        You should not need to override this function, so long as your sensor's
+        implementation of `thread_function()` has a way to stop when the
+        robot's `sensors_active` attribute becomes `False` on deletion.
```

Comment:
Might be preferable to have `sensor_active` be a state that's tied to the sensor itself? One less robot state variable to rely on and more control over each of the sensors' threads, and would remove the reliance on checking on the robot's state from Sensors that might not care.

Reply:
Yes, good idea!